### PR TITLE
feat: add initial support for arm builds

### PR DESCRIPTION
**Pull request by johanneskoester**
Support for arm builds (macos-M1 and linux-graviton) via CircleCI.

# TODOs

* [x] Merge aarch64 support in bioconda-utils: https://github.com/bioconda/bioconda-utils/pull/866
* [x] Support for downloading artifacts from CircleCI. This needs to be done [here](https://github.com/bioconda/bioconda-utils/blob/master/bioconda_utils/artifacts.py), analogous to azure, and switchable between azure and circleci via a CLI flag [here](https://github.com/bioconda/bioconda-utils/blob/1009e8c3dd26118cdb858ce3ca5b29baaf4f2f70/bioconda_utils/cli.py#L521). Does anybody volunteer for this (thanks a lot, I am very busy with teaching and Snakemake 8 at the moment)?
* [x] https://github.com/bioconda/bioconda-containers/pull/55 needs to be fixed and merged (@martin-g is on it).
* [x] Decide how this should be enabled. Activate it for all, and let people skip failing stuff via a selector (`# [arm64]`) on build::skip as usual, or shall we not build by default and only enable it if something special is added to extra? Decision of bioconda/core: opt-in in the first round, via a special field in extra (see next item)
* [x] add support in bioconda-utils to run on linux-arm if the following is added to the recipe:
```
extra:
  additional-platforms:
    - linux-aarch64
```
* [x] Update the documentation (see https://github.com/bioconda/bioconda-docs/pull/16)
* [x] Implement bulk updates on circle CI for arm. This requires porting .github/workflows/bulk.yml to circleci.

### CODE REVIEW

**Comment by Yikun** [comment]
Wow! Greate step! We will join and work together to make it happen!

Also cc @martin-g

**Comment by Yikun** [comment]
1. Added multiarch for conda installation: https://github.com/bioconda/bioconda-common/pull/33

2. Switch `MINICONDA_VER` to upper version
- Currently we are using [MINICONDA_VER=py37_4.8.3](https://github.com/bioconda/bioconda-common/blob/6c882f7687d8aee9cb8cd4100766512a51c47386/common.sh#L1).
- Linux-aarch64 supported conda since `py37_4.9.2`
- osx-arm64 supported conda since `py38_4.10.1`
- So, which version should we switch?

3. bioconda-utils support:  https://github.com/bioconda/bioconda-utils/pull/866

After above 3 changes, this PR should be green.

**Comment by Yikun** [comment]
Upgrade miniconda version to py37_4.9.2 in bioconda-common https://github.com/bioconda/bioconda-common/pull/34

Frankly, I'm not sure should we also upgrade to `py38_4.10.1` to also support osx-arm64, looks like bump version from `py37_4.8.3` to `py37_4.9.2` is a small forward step which has less impact.

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+      - restore_cache:
+          keys:
+            - bioconda-utils-{{ arch }}-{{ checksum "install-and-set-up-conda.sh" }}
+
+      - run:
+          name: Install bioconda-utils
+          command: |
+            sudo mkdir -p /opt/miniconda
+            sudo chown -R $USER /opt/miniconda
+            bash install-and-set-up-conda.sh
```
Comment (Yikun) [inline .circleci/config.yml:36]:
```
ERROR: File or directory already exists: '/opt/miniconda'
If you want to update an existing installation, use the -u option.

Exited with code exit status 1
```
[1] https://app.circleci.com/pipelines/github/Yikun/bioconda-recipes/11/workflows/42683084-4273-4ec5-8f25-f5d6f1b6be29/jobs/17

I hitted above issue, and resolved by change [install-and-set-up-conda.sh#L42](https://github.com/bioconda/bioconda-common/blob/e03cb8697f45928e918a029429d96befc352e2b1/install-and-set-up-conda.sh#L42):
```patch
- bash miniconda.sh -b -p "${MINICONDA_INSTALLATION_DIR}"
+ bash miniconda.sh -u -b -p "${MINICONDA_INSTALLATION_DIR}"
```

After this change, it works!

Comment (Yikun) [inline .circleci/config.yml:36]:
```suggestion
            sudo mkdir -p /opt/
            sudo chown -R $USER /opt/
            bash install-and-set-up-conda.sh
```

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+      - save_cache:
+          key: bioconda-utils-{{ arch }}-{{ checksum "install-and-set-up-conda.sh" }}
+          paths:
+            - /opt/miniconda
+
+      - run:
+          name: Build and test
+          command: |
+            conda activate bioconda-utils
```
Comment (Yikun) [inline .circleci/config.yml:46]:
```suggestion
            source /opt/mambaforge/etc/profile.d/conda.sh
            source /opt/mambaforge/etc/profile.d/mamba.sh
            mamba activate bioconda
```

File: .circleci/config.yml [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+          name: Install bioconda-utils
+          command: |
+            sudo mkdir -p /opt/miniconda
+            sudo chown -R $USER /opt/miniconda
+            bash install-and-set-up-conda.sh
+
+      - save_cache:
+          key: bioconda-utils-{{ arch }}-{{ checksum "install-and-set-up-conda.sh" }}
+          paths:
+            - /opt/miniconda
```
Comment (Yikun) [inline .circleci/config.yml:41]:
```suggestion
            - /opt/mambaforge/
```

File: .circleci/config.yml [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+          command: |
+            conda activate bioconda-utils
+            bioconda-utils build recipes config.yml --git-range origin/master HEAD
+
+      - run:
+          name: Prepare artifacts
+          command: |
+            (
+            mkdir -p /tmp/artifacts/packages
+            cd /opt/miniconda/envs/bioconda/conda-bld || exit 0
```
Comment (Yikun) [inline .circleci/config.yml:54]:
```suggestion
            cd /opt/mambaforge/envs/bioconda/conda-bld || exit 0
```

File: .circleci/config.yml [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+      - restore_cache:
+          keys:
+            - bioconda-utils-{{ arch }}-{{ checksum "install-and-set-up-conda.sh" }}
+
+      - run:
+          name: Install bioconda-utils
+          command: |
+            sudo mkdir -p /opt/miniconda
+            chown -R $USER:$USER /opt/miniconda
```
Comment (Yikun) [inline .circleci/config.yml:84]:
```suggestion
            sudo mkdir -p /opt
            chown -R $USER:$USER /opt
```

File: .circleci/config.yml [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+            bash install-and-set-up-conda.sh
+
+      - save_cache:
+          key: bioconda-utils-{{ arch }}-{{ checksum "install-and-set-up-conda.sh" }}
+          paths:
+            - /opt/miniconda
```
Comment (Yikun) [inline .circleci/config.yml:90]:
```suggestion
            - /opt/mambaforge
```

File: .circleci/config.yml [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+      - run:
+          name: Build and push
+          command: |
+            conda activate bioconda-utils
```
Comment (Yikun) [inline .circleci/config.yml:95]:
```suggestion
            source /opt/mambaforge/etc/profile.d/conda.sh
            source /opt/mambaforge/etc/profile.d/mamba.sh
            mamba activate bioconda
```

**Review by Yikun — COMMENTED** [review]
I test it locally in Linux aarch64 with latest bioconda-utils, it works! And the `config.yaml` should fix for latest bioconda-utils.

**Comment by Yikun** [comment]
FYI, @johanneskoester this also what we should be ready before we support aarch64: https://github.com/bioconda/bioconda-containers/pull/55

File: .circleci/config.yml [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+version: 2.1
+
+executors:
+  osx-arm64:
+    macos:
+      xcode: 14.2.0 # indicate your selected version of Xcode
+    resource_class: macos.m1.large.gen1
+  linux-aarch64:
+    machine:
+      image: ubuntu-2004:current
```
Comment (martin-g) [inline .circleci/config.yml:11]: It might be useful: Since recently it is possible to use Docker executor with ARM64 resources - https://circleci.com/docs/using-docker/#arm
This way you could use a custom Docker image that has most of the dependencies preinstalled. It will save a lot of time for the runs/builds!

Here is a small example:
- https://github.com/martin-g/circleci-arm64-test/blob/master/.circleci/config.yml#L17-L20
- https://app.circleci.com/pipelines/github/martin-g/circleci-arm64-test
Here I use `openEuler` Linux. CircleCI provides only Ubuntu images for the `machine` executor

**Comment by ewels** [comment]
Hi all! Just wanted to pop my head up and say that we (@SeqeraLabs and @nextflow-io / @nf-core communities) would love to help out getting ARM bioconda builds rolling.

I'm seeing a few semi-private efforts starting to nudge forward (eg. [Oxford Nanopore](https://labs.epi2me.io/inkling/), some Seqera [proof-of-concept testing](https://anaconda.org/seqera/repo)) and I'm a bit anxious that without bioconda to unite behind, we will start to see some fragmentation & duplication of efforts. 

Is there anything that we can do to move the bioconda infrastructure piece forward? I feel like we're coming to a place where we might be able to rally a fair amount of resources for this project, so if there's anywhere where that's needed then please say 🙏🏻 

Once the bioconda back-end is in place, we could also try to help out with any community efforts to crowdsource arm recipe work via hackathons / events. That could be fun :)

**Comment by martin-g** [comment]
@ewels Thank you for offering your help here!
I am not a member of the Bioconda team! I am also a user who tries to help with this task.

You might be interested in the progress so far. Please check:
- https://github.com/bioconda/bioconda-containers/pull/55
- https://github.com/bioconda/bioconda-recipes/issues/23454#issuecomment-1521574446

It would be great if you have time to test the proposed changes and either propose improvements or approve/vote for them!

**Comment by Yikun** [comment]
@ewels Great to hear about that!

BTW, we also have completed about 500+ bioconda (bioconductor & conda) packages validation based on [latest bioconda-utils](https://github.com/bioconda/bioconda-utils/releases/tag/v2.4.0)[1][2][3]. It really turns out that bioconda on ARM is possible.

We really need suggestion from Bioconda team about how we could move it forward in upstream.

[1] https://github.com/bioconda/bioconda-utils/pull/866
[2] https://github.com/bioconda/bioconda-utils/pull/867
[3] https://github.com/bioconda/bioconda-utils/pull/906

**Comment by ewels** [comment]
Wow, lots of discussions going on all over the place - more than I realised. Thanks both! And impressive work @Yikun!

It seems to me that although there is still some work to do / some PRs to be merged to support this on the back-end, _most_ things are about ready. Scanning over this, and chatting to @rpetit3, it sounds like the main thing missing is the CI infrastructure to run the automated builds. Is that approximately correct?

@dpryan79, over in [this comment](https://github.com/bioconda/bioconda-utils/issues/706#issuecomment-1061207620) you mentioned discussions with AWS about credits. Is anyone in @bioconda/core aware of progress on this front, or more generally aware of the current state of discussions around build infrastructure + funding for ARM packages? It's possible that I might be in a position to help to un-stick some of this, depending on what's needed. Happy to help out if I can be of any use.

File: .circleci/config.yml [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+    parameters:
+      os:
+        type: executor
+    executor: << parameters.os >>
+    steps:
+      - checkout
+
+      - run:
+          name: Fetch bioconda install script
+          command: wget https://raw.githubusercontent.com/bioconda/bioconda-common/master/install-and-set-up-conda.sh
```
Comment (martin-g) [inline .circleci/config.yml:25]:
```suggestion
          command: |
             wget https://raw.githubusercontent.com/bioconda/bioconda-common/master/install-and-set-up-conda.sh
             wget https://raw.githubusercontent.com/bioconda/bioconda-common/master/configure-conda.sh
```

**Comment by ewels** [comment]
Sounds great! Yes AWS credits are always a little scary and do need keeping an eye on. We've been using them heavily for @nf_core stuff for several years now and it's been (mostly) smooth sailing. Just need billing alerts and a handful of people to share responsibility for regularly checking remaining credits. I've been looking into methods to better automate some of this - eg. automated slack messages with remaining credit levels. Nothing to report yet, but there may be some hope yet.

> However, if I am not wrong, at least the non-macos stuff is free to use for us already. So we could get started with that, and then, with Seqera behind us, we might be able to finally get an answer from CircleCI on the M1 question?

Sounds great! I'll see if I can get anything moving in CircleCI circles and in the mean time let's crack on! 🎉

**Comment by johanneskoester** [comment]
Ok, seems to work!
There are four remaining blockers for this PR now:
1. Support for downloading artifacts from CircleCI. This needs to be done [here](https://github.com/bioconda/bioconda-utils/blob/master/bioconda_utils/artifacts.py), analogous to azure, and switchable between azure and circleci via a CLI flag [here](https://github.com/bioconda/bioconda-utils/blob/1009e8c3dd26118cdb858ce3ca5b29baaf4f2f70/bioconda_utils/cli.py#L521). Does anybody volunteer for this (thanks a lot, I am very busy with teaching and Snakemake 8 at the moment)?
2. https://github.com/bioconda/bioconda-containers/pull/55 needs to be fixed and merged (@martin-g is on it).
3. Decide how this should be enabled. Activate it for all, and let people skip failing stuff via a selector (`# [arm64]`) on build::skip as usual, or shall we not build by default and only enable it if something special is added to extra?
4. Update the documentation (again, happy to take a PR on https://github.com/bioconda/bioconda-docs for the same reason as item 1).

**Comment by aliciaaevans** [comment]
@johanneskoester I can start working on item 1 (artifacts from CircleCI).

**Comment by johanneskoester** [comment]
> @johanneskoester I can start working on item 1 (artifacts from CircleCI).

Thanks a lot @aliciaaevans, and welcome to the team!

File: .circleci/config.yml [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+            bioconda-utils build recipes config.yml --git-range origin/master HEAD
+
+      - run:
+          name: Prepare artifacts
+          command: |
+            (
+            mkdir -p /tmp/artifacts/packages
+            cd /opt/mambaforge/envs/bioconda/conda-bld || exit 0
+            find -name .cache | xargs rm -rf || true
+            for n in index.html channeldata.json linux-64 osx-64 noarch; do
```
Comment (aliciaaevans) [inline .circleci/config.yml:73]: @johanneskoester I'm not seeing the `linux-aarch64` artifacts in CircleCI, only json and html files for `noarch`. I am think this is probably where it is missing.

```suggestion
            for n in index.html channeldata.json linux-aarch64 linux-64 osx-64 noarch; do
```

Comment (aliciaaevans) [reply .circleci/config.yml:73]: Since it's minor, I'm going to got ahead and commit this change to see if it produces the artifacts so I can test the artifact download.

**Comment by Yikun** [comment]
> add support in bioconda-utils to run on linux-arm if the following is added to the recipe:

https://github.com/bioconda/bioconda-utils/pull/923 I drafted an initial PR for this task, feel free to review! Thanks

**Comment by aliciaaevans** [comment]
For branches where there's no `additional-platforms` enabled, it seems like a bit of a waste to do all the initial setup steps in CircleCI and then not end up building anything. I'm not sure if there's a clean way to skip or abort the job, though, and it would require checking the meta.yaml file(s) without using bioconda-utils (since it would not yet be installed). I played around a bit with using `yg` but it can't parse jinja templates.

Any thoughts on if that is worth pursuing?

**Comment by martin-g** [comment]
I think it is a good optimization!
Maybe you can use `sed` first to remove the jinja placeholders and then pipe it to `yq` ?
Something like: `cat $recipe/meta.yaml | sed -i 's/^(.*){{(.*)}}(.*)$/$1$2$3/g' | yq ...` (I am not sure about the syntax, it is Perl-ish !)

**Comment by Yikun** [comment]
```bash
$ grep -A 2 "additional-platforms:" recipes/abra2/meta.yaml
  additional-platforms:
    - linux-aarch64

$ grep -A 2 "additional-platforms:" recipes/abra2/meta.yaml | grep -q linux-aarch64
$ echo $?
0

$ grep -A 2 "additional-platforms:" recipes/abra2/meta.yaml | grep -q linux-x86
$ echo $?
1
```

A simple way without bioconda-utils install came to my mind might like above, even though it's not a perfect way, but seems a possible alternative workaround.

**Comment by aliciaaevans** [comment]
See https://github.com/bioconda/bioconda-recipes/pull/43708 for doing a check for any `additional-platforms` before installing bioconda-utils. (PR would merge to this branch, but if this gets merged to master first, I can update it.)

I'm going to start working on porting .github/workflows/bulk.yml to CircleCI next, assuming no one else has that in progress.

**Comment by martin-g** [comment]
> * [Make base images multiarch bioconda-containers#55](https://github.com/bioconda/bioconda-containers/pull/55) needs to be fixed and merged (@martin-g is on it).

This has been finished by @mbargull with:
- https://github.com/bioconda/bioconda-containers/pull/58
- https://github.com/bioconda/bioconda-containers/pull/59
- https://github.com/bioconda/bioconda-containers/pull/60

I also updated my script that uses the official images - https://github.com/bioconda/bioconda-recipes/issues/23454#issuecomment-1770590292

**Comment by milot-mirdita** [comment]
I updated the MMseqs2 recipe to compile on ARM if it helps testing:
https://github.com/bioconda/bioconda-recipes/pull/43976

Once this is merged I will update all the other recipes I maintain.

**Comment by Yikun** [comment]
@aliciaaevans Any info or problem about `bulk.yml` CircleCI porting? Pls let us know if we can help!

**Comment by aliciaaevans** [comment]
Thanks, @Yikun. I made some progress (https://github.com/bioconda/bioconda-recipes/pull/43995) but I've been sidetracked working through some issues with updating the bioconda-utils version (not directly related to ARM, but may hold things up).

**Comment by aliciaaevans** [comment]
Got a draft PR going for documentation updates: https://github.com/bioconda/bioconda-docs/pull/16 

Commits and/or PRs into that branch are welcome!

**Comment by daler** [comment]
Docs over at https://github.com/bioconda/bioconda-docs/pull/16 are mostly ready (thanks @bgruening @Yikun @martin-g for the reviews). This is the last item on the checklist above, but I think there might be a few more things to take care of.

I think it would be good to include container uploads to Biocontainers when we first roll this out. That way, we don't have to go back and find which packages are missing containers and build them later. It will also ensure that we have all of the moving parts working correctly while all of this is loaded up in our collective mental RAM.

Most of the pieces are in place for this. But [`.circleci.yml`](https://github.com/bioconda/bioconda-recipes/blob/arm-builds/.circleci/config.yml) is currently inconsistent. Specifically, the regular build is using docker executor and not making containers to be stored as artifacts, while the bulk build *is* using machine executor and making containers...but not yet pushing them. I believe this is because the regular build part was written before we had the arm64 build container.

Furthermore, the various containers are not consistently named/tagged. The *conda image* and *base images* are currently being built in bioconda-containers. The way they are built means that we get quay.io tags for the container and version with child manifests for amd64 and arm64. See [here](https://quay.io/repository/bioconda/create-env?tab=tags) for example (use expanded view). In contrast, the *build image* is being built from bioconda-utils and has different image names rather than child manifests (see [here](https://quay.io/repository/bioconda/bioconda-utils-build-env-cos7-aarch64?tab=tags) for aarch64 and [here](https://quay.io/repository/bioconda/bioconda-utils-build-env-cos7?tab=tags) for amd64).

By the way, see notes [in docs PR](https://github.com/bioconda/bioconda-docs/pull/16/files#diff-1ad2f3c7a97856c94d15e74f315c3d4c871fe4b823d58abb5b6ae368b7e3fba4) for what these are and where these are used, if you need a refresher like I did...


I propose the following:

1. **Build images consistently.** Create the build image that is used by `bioconda-utils build --docker` the same way as the conda/base images used by `bioconda-utils buid --mulled-test`. This way, all amd64 and arm64 images will be built and stored the same way on quay.io. Ideally, the build image should be removed from bioconda-utils and maintained in bioconda-containers with all the other containers, again for consistency.
2. **Make circleci jobs consistent.** Use machine executor for both regular builds and bulk.
3. **Make containers during regular builds and store as artifacts** so they are picked up in the subsequent `bioconda-utils handle-merged-pr` command. Currently, only conda packages are stored as artifacts.
4. **Update the [comment bot](https://github.com/bioconda/bioconda-containers/blob/e75e4dddc2016eaaf81c336ef8f937595d81aae5/images/bot/src/bioconda_bot/comment.py#L50-L56)** to show aarch64 container artifacts.
5. **Push containers** as part of the bulk builds and in the `bioconda-utils handle-merge-pr` for regular builds.

Thoughts?

**Comment by bgruening** [comment]
> I think it would be good to include container uploads to Biocontainers when we first roll this out.

:+1: 

This makes all perfect sense and is so detailed, I love it. Thanks a lot @daler!

**Comment by daler** [comment]
Update....

> 2. Make circleci jobs consistent. Use machine executor for both regular builds and bulk
> 3. Make containers during regular builds and store as artifacts so they are picked up in the subsequent bioconda-utils handle-merged-pr command. Currently, only conda packages are stored as artifacts.

These are both done in #44447, now merged into this PR (thanks @aliciaaevans)

The first part of item 5, 

> 5. Push containers as part of the bulk builds ...

is now done in a combination of https://github.com/bioconda/bioconda-utils/pull/941  and https://github.com/bioconda/bioconda-recipes/pull/44457 (both merged)

---

Things that still need addressing:

> 1. Build images consistently

> 4. Update the [comment bot](https://github.com/bioconda/bioconda-containers/blob/e75e4dddc2016eaaf81c336ef8f937595d81aae5/images/bot/src/bioconda_bot/comment.py#L50-L56) to show aarch64 container artifacts.

and the latter part of item 5, which I haven't tested yet:

> 5. ... in the `bioconda-utils handle-merge-pr` for regular builds.

**Comment by daler** [comment]
Updates:

> 1. Build images consistently

Looking into this some more, the containers are built so differently that I think it will take a substantial documentation and cleanup effort to unify them. They work fine now, it's just convoluted and hard to follow/troubleshoot. Since this is not required for ARM builds **I propose we leave this the way it is for now.**

> 4. Update the [comment bot](https://github.com/bioconda/bioconda-containers/blob/e75e4dddc2016eaaf81c336ef8f937595d81aae5/images/bot/src/bioconda_bot/comment.py#L50-L56) to show aarch64 container artifacts.

Started to be addressed in https://github.com/bioconda/bioconda-containers/pull/61, but upon closer inspection there's a bit more work to be done to support the bot fetching/reporting CircleCI artifacts. This doesn't affect *building*, just *fetching* via `BiocondaBot please fetch artifacts` in a PR. **I propose leaving this broken for now** and work on fixing the artifact fetching later.

So the last remaining piece is:

> 5. ... in the bioconda-utils handle-merge-pr for regular builds.

This is currently being tested in https://github.com/bioconda/bioconda-recipes/pull/44462. When that PR is merged into master, we need to see if the artifacts are correctly picked up across circleci and azure, uploaded to the channel and biocontainers repo, the container is set to public, and package and containers can be downloaded and used correctly. If all that works, then we are in good shape.

**Comment by martin-g** [comment]
I am not sure one can export `GITHUB_TOKEN` to external users, like CircleCI in this case.
https://docs.github.com/en/actions/security-guides/automatic-token-authentication - `GITHUB_TOKEN` is automatically created for each Github Action workflow run.
If I am right then you will need to create a custom token with the needed permissions and use it instead.

**Comment by martin-g** [comment]
> * Update the documentation (again, happy to take a PR on https://github.com/bioconda/bioconda-docs for the same reason as above).

@johanneskoester (or anyone with write permissions) Could you please add a link to https://github.com/bioconda/bioconda-docs/pull/16 for this TODO item in the issue [description](https://github.com/bioconda/bioconda-recipes/pull/40550#issue-1678597872)

**Comment by martin-g** [comment]
In case you use Visual Studio Code: CircleCI just announced a plugin for it - https://circleci.com/product/vscode-extension/
https://marketplace.visualstudio.com/items?itemName=circleci.circleci

**Comment by daler** [comment]
_(Ugh, should have done the merge from master in the GH interface, not the local rebase and push. Sorry for the commit noise)_

Anyway, to report back on all of this: CircleCI tokens are fixed and packages can be successfully uploaded to Anaconda. Images can also be uploaded to quay.io/biocontainers too, so artifacts are all working, **BUT** handling the images remains challenging and may need more discussion. Here's why (see details):

<details>
<summary>LOTS OF DETAILS...</summary>

We had initially discussed putting a `-aarch64` suffix on tags, but thought that using manifests would be much better (Recall that a manifest is a list of images, each on different arch, that is pulled automatically for the right arch when using a single tag). However, manifests will be problematic, because the images need to be collected from two different CI platforms and put into a single manifest for uploading. We had previously discussed checking for the existence of a manifest, downloading it (and image) if it exists, and add the local image to the manifest before pushing both images and manifest back. But this still has a race condition: if no manifest yet exists, both CI platforms may try pushing their single-image manifests. I verified that pushing a manifest completely replaces the existing, so in a race condition the latest one wins and we don't get both archs.I

Another approach would be to communicate across CI platforms with [workflow dispatch](https://docs.github.com/en/rest/actions/workflows?apiVersion=2022-11-28#create-a-workflow-dispatch-event):
- trigger a dummy GH Actions workflow from CircleCI once it has stored an artifact
- trigger another dummy GH Actions workflow from Azure.
- a downstream GH Actions "upload containers" job would depend on those two GH Actions jobs (see e.g. [this SO](https://stackoverflow.com/a/71489231)), collect artifacts across CircleCI and Azure, build a single manifest with both images, and push to quay.io.

This latter approach will still break when a PR is merged but artifacts are no longer available. `bioconda-utils` does have a `--fallback build` option, but that would need to be somehow changed to trigger new jobs on CircleCI and Azure, and trigger the above steps again. I haven't thought this part through carefully yet. But a cross-CI-platform DAG of workflow runs seems complex and brittle and makes me a bit nervous...

Another idea: maybe we could do a scheduled nightly job that, upon seeing multiple archs for a package in the conda channel, builds  multi-arch images out of those packages. I think with e.g. `--platform=linux/arm64` args to docker/buildah we could do this on the same platform, like GH Actions. Then upload the manifest and images. This could be a way of "patching" missing containers. This will need some more bioconda-utils and mulled work.

If we go with `-arch` suffixes, I think it would be straightforward to "patch" images later as well.

</details>

TL;DR: I propose:

1. **Merge this PR** to move forward with the linux-aarch64 conda package builds.
2. Revisit adding `arch` suffixes on biocontainer tags.
3. Discuss other ideas for collecting images from multiple CI platforms into a single manifest that can be pushed.
4. Patch missing arm biocontainer images later (since we'll need a solution for that situation no matter what)

**Comment by daler** [comment]
For item 2, yes, this is for the package images pushed to quay.io/biocontainers. Since manifests are a part of the [OCI spec](https://github.com/opencontainers/image-spec/blob/main/manifest.md) it's unlikely users would have trouble with it; in fact it's how many base docker images [e.g. alpine](https://hub.docker.com/_/alpine/tags) are already provided. It's nice because you can use the same `pull` command independent of architecture.

For item 4, the "patching" I had in mind was for cases where we had an arm package but not an arm container. So `extra-platforms: linux-aarch64` would already be there in the recipes. Once package containers are working, yep, a build number bump should then work, but it would still leave an "orphaned" older build number with no container. We'd need to decide whether that matters.

**Comment by martin-g** [comment]
> 3\. Discuss other ideas for collecting images from multiple CI platforms into a single manifest that can be pushed.

Would https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#check_suite work here ? 
The idea is to be notified when all checks (including the Azure and CircleCI) finish and then pull the new images and create a manifest for them.

**Comment by martin-g** [comment]
I played with this idea and created a Github App https://github.com/martin-g/my-first-github-app/blob/main/index.js that listens on `check_suite.completed` event and just logs the JSON payload.
Unfortunately the payload returns
```
...
"status": "completed",
"conclusion": "success",
...
```
even when a CircleCI check fails - https://github.com/martin-g/circleci-arm64-test/pull/5

If a Github Action check (`Pull Request / pull_request `) fails then the payload contains:
```
"status": "completed",
"conclusion": "failure",
```

**Comment by martin-g** [comment]
https://discuss.circleci.com/t/configure-circleci-checks-to-trigger-github-actions-workflow/50435 gives me hope! Testing it!

**Comment by martin-g** [comment]
It does work !
Now if my CircleCI workflow fails then the `check_suite.completed` event says `"conclusion": "failure",`!
I guess something similar could be done for the Azure checks too.

If the payload has `"conclusion": "success"` then the Github App could trigger another Github Actions workflow that will create the manifest!

**Comment by Yikun** [comment]
> 1.Merge this PR to move forward with the linux-aarch64 conda package builds.

+1, very exciting to to see this happening!

> 4. Patch missing arm biocontainer images later (since we'll need a solution for that situation no matter what)

+1, I think it is reasonable, we could improve it step by step!

**Review by daler — APPROVED** [review]

**Comment by jmarshall** [comment]
I see with this merge a linux-aarch64 build has now appeared [for the test package abra2](https://anaconda.org/bioconda/abra2/files). :tada:

Does this mean it is time for interested maintainers to start adding `additional-platforms:`/`- linux-aarch64` to individual recipes in ordinary PRs? Or is there more to do first? (Probably an update at #33333 would be useful.)

**Comment by daler** [comment]
Yes!! Please add `additional platforms: [linux-aarch64]` for recipes as you see fit 🚀

I'll update at #33333, need to get the docs fully deployed first though (that is in progress).

pysam, samtools, htslib are among the highest out-degree centrality of the entire bioconda channel DAG, so if you want to get those working that would be a huge help.

**Comment by jmarshall** [comment]
You will be unsurprised to hear those are precisely the packages I have in mind. :smile:
They are already built for ARM in their upstream CI, so I would not expect it to take much to get them working… 🤞

**Comment by rpetit3** [comment]
Awesome work everyone! Thank you @daler for taking the lead and everyone else for providing tremendous feedback. If you update recipes to opt-in to ARM builds, please feel free to request me as a reviewer!
